src/coreclr/jit/valuenum.cpp

```diff
@@ -8685,7 +8685,21 @@ void Compiler::fgValueNumberTree(GenTree* tree)
                         VNFunc loadFunc =
                             ((tree->gtFlags & GTF_IND_NONNULL) != 0) ? VNF_InvariantNonNullLoad : VNF_InvariantLoad;
 
-                        tree->gtVNPair = vnStore->VNPairForFunc(tree->TypeGet(), loadFunc, addrNvnp);
+                        // Special case: for initialized non-null 'static readonly' fields we want to keep field
+                        // sequence
+                        // to be able to fold their value
+                        ValueNum fieldSeqVN;
+                        if ((loadFunc == VNF_InvariantNonNullLoad) && addr->IsIconHandle(GTF_ICON_CONST_PTR))
+                        {
+                            fieldSeqVN = vnStore->VNForFieldSeq(addr->AsIntCon()->gtFieldSeq);
+                        }
+                        else
+                        {
+                            fieldSeqVN = vnStore->VNForNull();
```

Comment:
Field sequence VNs should only be obtained using `VNForFieldSeq`. `VNForNull` is not a valid field sequence VN.